eppo_core/src/ufc/rules.rs

```diff
@@ -0,0 +1,337 @@
+use regex::Regex;
+use semver::Version;
+
+use crate::{
+    ufc::{Condition, ConditionOperator, ConditionValue, Rule, Value},
+    AttributeValue, Attributes,
+};
+
+impl Rule {
+    pub(crate) fn eval(&self, attributes: &Attributes) -> bool {
+        self.conditions
+            .iter()
+            .all(|condition| condition.eval(attributes))
+    }
+}
+
+impl Condition {
+    fn eval(&self, attributes: &Attributes) -> bool {
+        self.operator
+            .eval(attributes.get(&self.attribute), &self.value)
+    }
+}
+
+impl ConditionOperator {
+    /// Applying `Operator` to the values. Returns `false` if the operator cannot be applied or
+    /// there's a misconfiguration.
+    fn eval(&self, attribute: Option<&AttributeValue>, condition_value: &ConditionValue) -> bool {
+        self.try_eval(attribute, condition_value).unwrap_or(false)
+    }
+
+    /// Try applying `Operator` to the values, returning `None` if the operator cannot be applied.
+    fn try_eval(
+        &self,
+        attribute: Option<&AttributeValue>,
+        condition_value: &ConditionValue,
+    ) -> Option<bool> {
+        match self {
+            Self::Matches | Self::NotMatches => {
+                let s = match attribute {
+                    Some(AttributeValue::String(s)) => s.clone(),
+                    Some(AttributeValue::Boolean(b)) => b.to_string(),
+                    _ => return None,
+                };
+                let regex = match condition_value {
+                    ConditionValue::Single(Value::String(s)) => Regex::new(s).ok()?,
+                    _ => return None,
+                };
+                let matches = regex.is_match(&s);
+                Some(if matches!(self, Self::Matches) {
+                    matches
+                } else {
+                    !matches
+                })
+            }
+
+            Self::OneOf | Self::NotOneOf => {
+                let s = match attribute {
+                    Some(AttributeValue::String(s)) => s.clone(),
+                    Some(AttributeValue::Number(n)) => n.to_string(),
+                    Some(AttributeValue::Boolean(b)) => b.to_string(),
```

Comment:
fancy--so we're trying toString() on all possible interpretations of the attribute value?

Reply:
not "trying" per-se: the code branch executes whatever type `attribute` is like a `switch` statement.